omero_certificates/certificates.py

```diff
@@ -45,81 +56,82 @@ def set_if_empty(cfgkey, default):
+        with open(keypath, "rb") as pem_openssl_key:
+            rsa_private_key = serialization.load_pem_private_key(
+                pem_openssl_key.read(),
+                password=None,
+            )
     else:
         log.info("Creating self-signed CA key: %s", keypath)
-        run_openssl(["genrsa", "-out", keypath, "2048"])
+        # Do what `openssl genrsa -out <keypath> <numbits>` would do
+        rsa_private_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
+        with open(keypath, "wb") as pem_openssl_key:
+            pem_openssl_key.write(
+                rsa_private_key.private_bytes(
+                    Encoding.PEM,
+                    PrivateFormat.TraditionalOpenSSL,  # Essentially PKCS#1
+                    NoEncryption(),
+                )
+            )
         created_files.append(keypath)
 
     # Self-signed certificate
     log.info("Creating self-signed certificate: %s", certpath)
-    run_openssl(
-        [
-            "req",
-            "-new",
-            "-x509",
-            "-subj",
-            "{}/CN={}".format(owner, cn),
-            "-days",
-            days,
-            "-key",
-            keypath,
-            "-out",
-            certpath,
-            "-extensions",
-            "v3_ca",
-        ]
+    # Do what `openssl req -x509 ...` would do
+    utcnow = datetime.utcnow()
+    try:
+        subject = issuer = x509.Name.from_rfc4514_string("{},CN={}".format(owner, cn))
+    except ValueError:
+        return (
+            f"'omero.certificates.owner' configuration setting '{owner}' not a "
+            "valid RFC 4514 string!  Are you upgrading?  See "
+            "https://pypi.org/project/omero-certificates/ for help."
+        )
+    cert = (
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(issuer)
+        .not_valid_before(utcnow)
+        .not_valid_after(utcnow + timedelta(days=days))
+        .public_key(rsa_private_key.public_key())
+        .serial_number(x509.random_serial_number())
+        .sign(rsa_private_key, SHA256())
     )
+    with open(certpath, "wb") as pem_cert:
+        pem_cert.write(cert.public_bytes(Encoding.PEM))
     created_files.append(certpath)
 
     # PKCS12 format
     log.info("Creating PKCS12 bundle: %s", pkcs12path)
-    run_openssl(
-        [
-            "pkcs12",
-            "-export",
-            "-out",
-            pkcs12path,
-            "-inkey",
-            keypath,
-            "-in",
-            certpath,
-            "-name",
-            "server",
-            "-password",
-            "pass:{}".format(password),
-        ]
-    )
+    # Do what `openssl pkcs12 ...` would do
+    with open(pkcs12path, "wb") as p12:
+        p12.write(
+            pkcs12.serialize_key_and_certificates(
+                b"server",
+                rsa_private_key,
+                cert,
+                None,
+                BestAvailableEncryption(password.encode("utf-8")),
```

Comment:
Tested on OSX, and the default version broke the server connection with the following relevant line in `master.err`
```
   plug-in initialization failed: IceSSL: error reading certificate `/Users/sbesson/OMERO.repo/certs/server.p12':
   (error: -25264 description: MAC verification failed during PKCS12 import (wrong password?))
```
As explained in https://openradar.appspot.com/FB8988319 and https://cryptography.io/en/latest/hazmat/primitives/asymmetric/serialization/#cryptography.hazmat.primitives.serialization.pkcs12.serialize_key_and_certificates, this is a compatibility issue which affects OS X and Windows. To start the server with the new self-signed SSL certificate, , I had to apply to use a less secure encryption as described in the API docs with the following patch

```diff
(base) sbesson@Sebastiens-MacBook-Pro-2 omero-certificates % diff /opt/miniconda3/envs/omeroserver/lib/python3.8/site-packages/omero_certificates/certificates.py omero_certificates/certificates.py 
13c13
< from cryptography.hazmat.primitives.hashes import SHA256, SHA1
---
> from cryptography.hazmat.primitives.hashes import SHA256
126,131d125
<         encryption = (
<             PrivateFormat.PKCS12.encryption_builder().
<             kdf_rounds(50000).
<             key_cert_algorithm(pkcs12.PBES.PBESv1SHA1And3KeyTripleDESCBC).
<             hmac_hash(SHA1()).build(password.encode("utf-8"))
<         )
138c132
<                 encryption,
---
>                 BestAvailableEncryption(password.encode("utf-8")),
````

I believe it's a server-side issue i.e. using the `BestAvailableEncryption` on a Linux environment should not affect the ability to connect from OSX or Windows but this should definitely be tested.
If this is the case, this raises once more the question of whether we want to keep supporting server deployment on OSX moving forward (and include the necessary compatibility adjustments) or restrict the deployment options.